### ISSUE TITLE
`--update` flag causes sync command to skip checksum checks

### ISSUE THREAD

**Issue by cfbao** [rclone/rclone#3232]
#### What is the problem you are having with rclone?
By default, `rclone sync` will
1. check file sizes and modtimes
2. if there's a difference in modtime, check checksums
3. if checksums are different, transfer from source to destination.

When used with the `--update` flag, I expect the process to be similar, except it'd only proceed beyond the 1st step if the modtime at the destination is older (but not newer) than the modtime at the source.

However, what `rclone sync --update` actually does is:
1. check file sizes and modtimes
2. if modtime at destination is older, directly proceed to transfer from source to destination without checking checksums

#### What is your rclone version (output from `rclone version`)
```
rclone v1.47.0-086-ge7c20e0b-beta
- os/arch: windows/amd64
- go version: go1.12.3
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Windows 7 Pro 64-bit

####  Which cloud storage system are you using? (eg Google Drive)
First discovered when using Google Drive; reproduced on local filesystem.

#### The command you were trying to run
Discovered the problem when running
`rclone sync src dst --update -v`

For better clarity, here's a more elaborated sequence of commands
```
> rclone lsf src
test_file
> rclone check src dst
2019/05/28 11:45:52 NOTICE: Local file system at \\?\D:\test\dst: 0 differences found
2019/05/28 11:45:52 NOTICE: Local file system at \\?\D:\test\dst: 1 matching files
> rclone touch src/test_file
> rclone sync src dst --no-update-modtime --stats 0 -v
2019/05/28 11:48:49 INFO  : Local file system at \\?\D:\test\dst: Waiting for checks to finish
2019/05/28 11:48:49 INFO  : Local file system at \\?\D:\test\dst: Waiting for transfers to finish
2019/05/28 11:48:49 INFO  : Waiting for deletions to finish
> rclone sync src dst --update --no-update-modtime --stats 0 -vv
```

#### A log from the command with the `-vv` flag
log of the last command:
```
2019/05/28 12:03:53 DEBUG : rclone: Version "v1.47.0-086-ge7c20e0b-beta" starting with parameters ["rclone" "sync" "src" "dst" "--update" "--no-update-modtime" "--stats" "0" "-vv"]
2019/05/28 12:03:53 DEBUG : Using config file from "C:\\Users\\Chenfeng\\.config\\rclone\\rclone.conf"
2019/05/28 12:03:53 INFO  : Local file system at \\?\D:\test\dst: Waiting for checks to finish
2019/05/28 12:03:53 DEBUG : test_file: Destination is older than source, transferring
2019/05/28 12:03:53 INFO  : Local file system at \\?\D:\test\dst: Waiting for transfers to finish
2019/05/28 12:03:53 INFO  : test_file: Copied (replaced existing)
2019/05/28 12:03:53 INFO  : Waiting for deletions to finish
2019/05/28 12:03:53 DEBUG : 4 go routines active
2019/05/28 12:03:53 DEBUG : rclone: Version "v1.47.0-086-ge7c20e0b-beta" finishing with parameters ["rclone" "sync" "src" "dst" "--update" "--no-update-modtime" "--stats" "0" "-vv"]
```

**Comment by ncw** [comment]
This feature is modeled on rsync.

Here is the rsync man page

```
      -u, --update
              This  forces rsync to skip any files which exist on the destina‐
              tion and have a modified time that  is  newer  than  the  source
              file.   (If an existing destination file has a modification time
              equal to the source file’s, it will be updated if the sizes  are
              different.)

              Note that this does not affect the copying of dirs, symlinks, or
              other special files.  Also, a difference of file format  between
              the  sender  and  receiver  is always considered to be important
              enough for an update, no matter what date is on the objects.  In
              other words, if the source has a directory where the destination
              has a file, the transfer would occur  regardless  of  the  time‐
              stamps.

              This  option  is  a transfer rule, not an exclude, so it doesn’t
              affect the data that goes  into  the  file-lists,  and  thus  it
              doesn’t affect deletions.  It just limits the files that the re‐
              ceiver requests to be transferred.
```

Can you see what rsync does in this case?

**Comment by cfbao** [comment]
I'm not very familiar with rsync, but from what I gathered, rsync doesn't do checksum comparison even in default mode, so it makes sense that its "update" mode doesn't do checksum comparison either.

I interpret the man page to mean that "update" mode should only reduce data transfer, not increase.
In rclone's case, because it already does checksum comparison before transfer in default mode, it makes more sense that rclone should also do checksum comparison in update mode.

**Comment by ncw** [comment]
This is what the rclone docs say about [-u, --update](https://rclone.org/docs/#u-update):

> This forces rclone to skip any files which exist on the destination and have a modified time that is newer than the source file.
> 
> If an existing destination file has a modification time equal (within the computed modify window precision) to the source file’s, it will be updated if the sizes are different.

So in pseudocode

- if dest modification time newer than source - **skip**
- if dest modification time older than source - **transfer**
- if dest modification time equal to source
    - if sizes different - **transfer**
    - if sizes the same - **skip**

I think this is what you are proposing - is that correct?

- if dest modification time newer than source - **skip**
- if dest modification time older than source - **transfer**
- if dest modification time equal to source
    - if sizes different - **transfer**
    - if sizes the same
        - if checksum not available - **transfer**
        - if checksums differ - **transfer**
        - otherwise **skip**

**Comment by cfbao** [comment]
Not exactly. What I have in mind is:
* if dest modification time newer than source - **skip**
* if dest modification time older than or equal to source
  * if sizes different - **transfer**
  * if sizes the same
    * if checksum not available - **transfer**
    * if checksums differ - **transfer**
    * otherwise **skip** (or maybe **update modtime**?)

In other words, the only difference between `sync --update` and `sync` (without `--update`) is that, at the beginning of the file comparison logic, the update mode skips files with newer modtimes at dest, but rest of the comparison logic is the same as the default sync mode.

At least, this is what I thought `sync --update` should do, and what I want personally. I also think this is what the rsync man page of `--update` means. When I read the docs, I thought `--update` would only
* introduce more checks based on modtimes, but not mess with checksum comparison
* reduce but never increase data transfer

Maybe I was mistaken, but I was rather surprised that a lot of unnecessary data transfer happened due to the `--update` flag.

**Comment by ncw** [comment]
I did a test with rsync.

From the below I gather that
- rsync transfers the file if the modtime is different (like rclone) without `-u`
- rsync transfer the file only if it is newer with `-u`

```
$ mkdir src
$ truncate -s 10M src/one.txt
$ cp -av src dst

$ touch src/one.txt 
$ rsync -avv src/ dst/
sending incremental file list
delta-transmission disabled for local transfer or --whole-file
one.txt
total: matches=0  hash_hits=0  false_alarms=0 data=10485760

sent 10,488,439 bytes  received 105 bytes  20,977,088.00 bytes/sec
total size is 10,485,760  speedup is 1.00

$ touch dst/one.txt 
$ rsync -avv src/ dst/
sending incremental file list
delta-transmission disabled for local transfer or --whole-file
one.txt
total: matches=0  hash_hits=0  false_alarms=0 data=10485760

sent 10,488,439 bytes  received 105 bytes  20,977,088.00 bytes/sec
total size is 10,485,760  speedup is 1.00

$ touch src/one.txt 
$ rsync -auvv src/ dst/
sending incremental file list
delta-transmission disabled for local transfer or --whole-file
one.txt
total: matches=0  hash_hits=0  false_alarms=0 data=10485760

sent 10,488,438 bytes  received 105 bytes  20,977,086.00 bytes/sec
total size is 10,485,760  speedup is 1.00

$ touch dst/one.txt 
$ rsync -auvv src/ dst/
sending incremental file list
delta-transmission disabled for local transfer or --whole-file
one.txt is newer
total: matches=0  hash_hits=0  false_alarms=0 data=0

sent 83 bytes  received 107 bytes  380.00 bytes/sec
total size is 10,485,760  speedup is 55,188.21
$
```

So as far as rsync is concerned your statement below is correct.  As part of the transfer rsync is doing the equivalent of checking the --checksum

> In other words, the only difference between sync --update and sync (without --update) is that, at the beginning of the file comparison logic, the update mode skips files with newer modtimes at dest, but rest of the comparison logic is the same as the default sync mode.

In rclone terms, your pseudo code would be equivalent to using `--checksum` but skipping newer files first.  I think that would be easy to explain at least!

So I'd propose something like this in the docs for `--update`

> This forces rclone to skip any files which exist on the destination and have a modified time that is newer than the source file.
>
> If an existing destination file has a modification time equal (within the computed modify window precision) or older than the source file’s, it will be updated if either of the size or checksum are different..

I had a go at implementing this here - can you give it a go and see if you think it works as you expect.

https://beta.rclone.org/branch/v1.47.0-108-g43e2f95b-fix-3232-update-beta/ (uploaded in 15-30 mins)

PS I just have a nagging suspicion that people rely on the current behavior but I can't find evidence for it at the moment!

**Comment by cfbao** [comment]
Oops! My pseudocode isn't actually what I want! Sorry!

> In other words, the only difference between sync --update and sync (without --update) is that, at the beginning of the file comparison logic, the update mode skips files with newer modtimes at dest, but rest of the comparison logic is the same as the default sync mode.

This description is still correct, but my pseudocode is wrong. What I had in mind is without the `--checksum` flag, i.e. the only effect of `--update` is to skip newer files first. Does such behaviour make sense?
If users want to check all checksums, the `--checksum` flag can still be added on top of `--update`.

The correct pseudocode that actually describes what I want is:

* if dest modification time newer than source - **skip**
* if dest modification time equal to source
  * if sizes different - **transfer**
  * if sizes the same - **skip**
* if dest modification time older than source
  * if sizes different - **transfer**
  * if sizes the same
    * if checksum not available - **transfer**
    * if checksums differ - **transfer**
    * otherwise - **update modtime**

**Comment by ncw** [comment]
> The correct pseudocode that actually describes what I want is:
> 
> * if dest modification time newer than source - **skip**
> * if dest modification time equal to source
>   
>   * if sizes different - **transfer**
>   * if sizes the same - **skip**
> * if dest modification time older than source
>   
>   * if sizes different - **transfer**
>   * if sizes the same
>     
>     * if checksum not available - **transfer**
>     * if checksums differ - **transfer**
>     * otherwise - **update modtime**

That is actually quite a lot closer to the original algorithm.

I've had a go at that here

https://beta.rclone.org/branch/v1.47.0-120-gbd3f2c64-fix-3232-update-2-beta/ (uploaded in 15-30 mins)

It doesn't update the modtime though as it is using the existing `--checksum` path which doesn't do that.

**Comment by cfbao** [comment]
The tests/build seem to have failed?

**Comment by ncw** [comment]
Sorry I messed that up and broke some tests.  Will have another go in a bit!

**Comment by ncw** [comment]
Sorry I dropped the ball on this!

Here is the latest code which implements this pseudo code

https://beta.rclone.org/branch/v1.49.5-197-ga007919d-fix-3232-update-2-beta/ (uploaded in 15-30 mins)

* if dest modification time newer than source - **skip**
* if dest modification time equal to source
  * if sizes different - **transfer**
  * if `---checksum` is set then and sums different then **transfer**
  * if sizes the same - **skip**
* if dest modification time older than source
  * if sizes different - **transfer**
  * if sizes the same
    * if checksum not available - **transfer**
    * if checksums differ - **transfer**
    * otherwise - **update modtime** (this is implemented now)

What do you think?

**Comment by cfbao** [comment]
You never need to apologize for your FOSS work!

I just tried it out. It looks great!
I understand that the "update modtime" part isn't yet included in this beta build, right?

**Comment by ncw** [comment]
> I just tried it out. It looks great!

Thanks

> I understand that the "update modtime" part isn't yet included in this beta build, right?

It should be - did it not work?

Though whether the update modtime works depends on which backend you use it on.

**Comment by cfbao** [comment]
> It should be - did it not work?

Not in my tests.

> Though whether the update modtime works depends on which backend you use it on.

I tried both local and Google Drive. Neither updated modtime.

```plain
>rclone lsl test
 16777216 2019-10-10 13:55:26.452649100 random

>rclone sync test test2

>rclone touch test\random

>rclone lsl test
 16777216 2019-10-10 14:02:09.440698700 random

>rclone lsl test2
 16777216 2019-10-10 13:55:26.452649100 random

>rclone sync -u -vv --stats 0 test test2
2019/10/10 14:02:47 DEBUG : rclone: Version "v1.49.5-197-ga007919d-fix-3232-update-2-beta" starting with parameters ["rclone" "sync" "-u" "-vv" "--stats" "0" "test" "test2"]
2019/10/10 14:02:47 DEBUG : Using config file from "C:\\Users\\Chenfeng\\My Docs\\Local\\rclone.conf"
2019/10/10 14:02:47 INFO  : Local file system at //?/C:/Users/Chenfeng/test2: Waiting for checks to finish
2019/10/10 14:02:47 DEBUG : random: MD5 = 4c59a34e3b174eb51ef53b4ffc6432db OK
2019/10/10 14:02:47 DEBUG : random: Size and MD5 of src and dst objects identical
2019/10/10 14:02:47 DEBUG : random: Unchanged skipping
2019/10/10 14:02:47 INFO  : Local file system at //?/C:/Users/Chenfeng/test2: Waiting for transfers to finish
2019/10/10 14:02:47 INFO  : Waiting for deletions to finish
2019/10/10 14:02:47 DEBUG : 4 go routines active
2019/10/10 14:02:47 DEBUG : rclone: Version "v1.49.5-197-ga007919d-fix-3232-update-2-beta" finishing with parameters ["rclone" "sync" "-u" "-vv" "--stats" "0" "test" "test2"]

>rclone lsl test
 16777216 2019-10-10 14:02:09.440698700 random

>rclone lsl test2
 16777216 2019-10-10 13:55:26.452649100 random
```

**Comment by ncw** [comment]
Ah, I see what is wrong!  Try this :-)

https://beta.rclone.org/branch/v1.49.5-203-g00b29e84-fix-3232-update-2-beta/ (uploaded in 15-30 mins)

**Comment by cfbao** [comment]
Just tested all scenarios I could think of, and it works flawlessly!
Thanks a lot!

**Comment by ncw** [comment]
Thanks for testing :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.50

**Comment by anhduong0811** [comment]
My syncflows like as:

```bash
#!/bin/bash

## Create a flag file that indicate the first time run after reboot/turn on
flag="/tmp/rcloneFlagFile"

## If there is no flag, this is the 1st time
if [[ ! -f "$flag" ]]; then
	touch "$flag"
	rclone bisync "mybox:Notes" /Notes --resync 
fi

## Sync regularly
if [[ -f "$flag" ]]; then
	rm "$flag"
	inotifywait -m -r -e close_write /Notes | while read path action file; do 
	sleep 12;
	rclone sync /Notes "mybox:Notes" -u --ignore-errors --track-renames --backup-dir "mybox:.rclone-backup/Notes" --suffix .$(date +"%Y-%m-%d-%H-%M-%S"); 
	sleep 9;
	rclone copy -u -v "mybox:Notes" /Notes --backup-dir /Notes.Backup --suffix .$(date +"%Y-%m-%d-%H-%M-%S"); 
done
fi
```

**Comment by nielash** [comment]
> * I think the problem here is because rclone is not a tool designed for two-way synchronization (to be exact, it is a one-way synchronization tool), so it does not keep a ledger to track files changes (modtime, size, checksum) in different places to compare and decide which to keep, which to discard or at least create duplicate files and notify the user.
> 
> I wonder if there is any solution to this problem?

@anhduong0811 This is exactly what [`rclone bisync`](https://rclone.org/bisync/) is designed for.

From your script example it looks like you are using `bisync` only the first time, and only with the `--resync` flag. Then after this, you use `sync` and `copy` instead of `bisync`. Is there a reason for doing that instead of just using `bisync` (without `--resync`) the way it's intended? I think it would help with the problems you mentioned.


> Sorry for bringing up this topic again. But I have a few lingering questions but don't know where else to ask.

Generally [the forum](https://forum.rclone.org/latest) is the best place 🙂

**Comment by anhduong0811** [comment]
> From your script example it looks like you are using `bisync` only the first time, and only with the `--resync` flag. Then after this, you use `sync` and `copy` instead of `bisync`. Is there a reason for doing that instead of just using `bisync` (without `--resync`) the way it's intended? I think it would help with the problems you mentioned.

`bisync` does not work well with `inotifywait`, I got "fatal error, need resync" almost at the time it runs, so I only use it as a first sync (with `--resync` flag). 

> Generally [the forum](https://forum.rclone.org/latest) is the best place 🙂
I didn't understand clearly the "--update" flag used with the `sync` command, so I tried searching Google first and found this topic.

I feel like it would be a bit troublesome just to ask a question, but I have to create a thread on the forum xd
(If rclone have irc channels, matrix or discord to ask, it would be more convenient.. but anyway, ask here so many people can find it when they need it)

**Comment by nielash** [comment]
> `bisync` does not work well with `inotifywait`, I got "fatal error, need resync" almost at the time it runs

There are some new features coming soon that should significantly improve this:

- https://github.com/rclone/rclone/pull/7501
- https://github.com/rclone/rclone/pull/7529
- https://github.com/rclone/rclone/pull/7410

As you correctly noted, `copy` and `sync` are one-way synchronization tools, and are not intended for two-way sync because of issues like the ones you described. I'd encourage you to give bisync another try once the new features are available -- I think it will be a better fit for your use case.

Happy to chat through your questions in more detail with you if you'd like (but let's do it on a new issue or in the forum, as this is an old closed issue from 2019.)

**Comment by anhduong0811** [comment]
> > `bisync` does not work well with `inotifywait`, I got "fatal error, need resync" almost at the time it runs
> 
> There are some new features coming soon that should significantly improve this:
> 
>     * [bisync: Graceful Shutdown, --recover from interruptions without --resync #7501](https://github.com/rclone/rclone/pull/7501)
> 
>     * [bisync: auto-resolve conflicts, customize `--resync` - fixes #7471 #5681 #7529](https://github.com/rclone/rclone/pull/7529)
> 
>     * [bisync: logger, march, and several big bug fixes #7410](https://github.com/rclone/rclone/pull/7410)
> 
> 
> As you correctly noted, `copy` and `sync` are one-way synchronization tools, and are not intended for two-way sync because of issues like the ones you described. I'd encourage you to give bisync another try once the new features are available -- I think it will be a better fit for your use case.

ah, it's nice to try them!!!
I will try and give it a feedback ^^

> Happy to chat through your questions in more detail with you if you'd like (but let's do it on a new issue or in the forum, as this is an old closed issue from 2019.)

Haha, yeah, I'm thinking so. This has reference value, but referencing it in a newer thread would be better.
But anyway, they are not new problems, it would be good if we could connect them... so that when we look at it, we can see the development process of the project, the challenges posed and how we overcome them :handshake:
